### ISSUE TITLE
False positive E0308 for macro_rules with once_cell::sync::Lazy when macro expansion is enabled

### ISSUE THREAD

**Issue by Obstbeeren** [intellij-rust/intellij-rust#10694]
## Environment

* **IntelliJ Rust plugin version:** 0.4.198.5409-231
* **Rust toolchain version:** 1.71.0 (8ede3aae2 2023-07-12) x86_64-unknown-linux-gnu
* **IDE name and version:** PyCharm 2023.1.4 Professional Edition (PY-231.9225.15)
* **Operating system:** Linux 6.4.2-3-manjaro
* **Macro expansion:** enabled
* **Additional experimental features:** org.rust.macros.proc.attr

## Problem description
False positive E0308 for macro_rules with once_cell::sync::Lazy. This is only when macro expansion is enabled.
```
mismatched types [E0308]
expected `Lazy<&str, fn() -> &str>`, found `Lazy<&str, fn() -> &mut str>` 
```
## Steps to reproduce

```rust
use once_cell::sync::Lazy;

#[macro_export]
macro_rules! get_env {
    ($name:expr) => {
        Lazy::new(|| {
            Box::leak(
                std::env::var($name)
                    .expect(&format!("Missing env {}", $name))
                    .into_boxed_str(),
            )
        })
    };
}

static SOME_ENV: Lazy<&'static str> = get_env!("SOME_ENV");
```
